src/script/sign.cpp

```diff
@@ -144,8 +144,10 @@ static bool SignStep(const SigningProvider& provider, const BaseSignatureCreator
         ret.push_back(valtype()); // workaround CHECKMULTISIG bug
         for (size_t i = 1; i < vSolutions.size() - 1; ++i) {
             CPubKey pubkey = CPubKey(vSolutions[i]);
-            if (ret.size() < required + 1 && CreateSig(creator, sigdata, provider, sig, pubkey, scriptPubKey, sigversion)) {
-                ret.push_back(std::move(sig));
+            if (CreateSig(creator, sigdata, provider, sig, pubkey, scriptPubKey, sigversion)) {
+                if (ret.size() < required + 1) {
```

Comment:
The nice thing about the way it was done before is that if this expr returns false it would get short circuited and `CreateSig` wouldn't be called too.

Now it's the opposite. which I think makes less sense when we know that CreateSig is more work than that simple expr, so it might do CreateSig just to fail that if afterwards.

Reply:
This is necessary for `LegacyScriptPubKeyMan::CanProvide` to pass. That uses the `DUMMY_SIGNATURE_CREATOR` which creates dummy signatures without checking whether the `SigningProvider` has the private keys (as it should). However, in `CanProvide`, we want to make sure that we do actually have the private keys, and if the dummy signing returns us dummy signatures that correspond to keys that we don't have, `CanProvide` will fail. So we need it to give us all of the signatures (by putting them in the `sigdata`) so that `CanProvide` can check for whether it actually could sign.

Reply:
I added a comment.

Reply:
I don't think it makes sense to have a comment with such specific information about the wallet in the non-wallet signing code. There's already a comment there explaining why we try to sign for all keys.